### DISCUSSION TITLE
We should have metrics about the performance of detekt

### DISCUSSION THREAD

**Discussion by BraisGabin** [detekt/detekt#5183] [Ideas]
The performance in a static analysis rule is really important and we are nearly blind about it. Which is our slower rule? Could we fix it? If we "improve the code" are we really improving it?

The performance is so critical that a lot of our users use plain detekt instead of type solving to get that speed bust. And we know that the rules that we have with type solving are *really* good.

My idea:

We could have a nighly task that runs detekt over some open source projects. We should track for each file how many time we expend on each rule. Save that in a data base and plot it using Grafana or something similar.

Ideas about how to do this? Do someone want to give it a try?

**Comment by schalkms** [comment]
I'm not sure about this. Measuring the performance is one thing, improving it is a whole different story, since detekt heavily depend on Kotlin compiler APIs.
The type solving task costs time, but only one alternative with the compiler-plugin exists. The compiler-plugin potentially speeds things up, but isn't in a state to be integrated in detekt IMHO.

**Comment by arturbosch** [comment]
> The performance in a static analysis rule is really important and we are nearly blind about it. Which is our slower rule? Could we fix it? If we "improve the code" are we really improving it?
> 
> The performance is so critical that a lot of our users use plain detekt instead of type solving to get that speed bust. And we know that the rules that we have with type solving are _really_ good.
> 
> My idea:
> 
> We could have a nighly task that runs detekt over some open source projects. We should track for each file how many time we expend on each rule. Save that in a data base and plot it using Grafana or something similar.
> 
> Ideas about how to do this? Do someone want to give it a try?

We do have `PerformanceMonitor` as a performance measuring tool for each phase like parsing, binding, analysis etc.
These stats should be printed when using the `--debug` flag.

We could use a similar approach to measure the performance of rules per `KtFile` and accumulate them to get the run time of a rule.

Following could be a good position to measure:

https://github.com/detekt/detekt/blob/0d42e58e1087ac5bffcc0a125d707efbd9bced88/detekt-core/src/main/kotlin/io/gitlab/arturbosch/detekt/core/Analyzer.kt#L121-L122

I also agree with @schalkms that the compiler plugin api would need to huge performance as we may skip parsing and binding phases.
However measuring rule performance is still a nice to have. This way we can find outlier rules which take for example 10 seconds instead of 1 like other rules (e.g. due to bad algorithm).

**Comment by eygraber** [comment]
Not saying that anything from here could be used considering they're different tools, but I've had great experiences with the android lint profiler https://github.com/google/android-lint-performance-probe

Maybe there's some inspiration there.

**Comment by mgroth0** [comment]
I am currently trying to figure out how to profile my Detekt analyses. Are there any updates on this? I want to get a break down of time spent per file and per rule.

**Reply by BraisGabin** [reply]
No updates but this is something that would be great to have. The main reason I didn't imminent it yet is because I have no idea how the output should look like. A value per `file` x `rule` would be huge. But that would allow to extract that information to other tools and perform different analysis. Something more simple would be easier to spot the main problems but it would hide a lot of information.

**Comment by mgroth0** [comment]
See my proposal https://github.com/detekt/detekt/issues/6962
